packages/remix-react/browser.tsx

```diff
@@ -381,6 +411,80 @@ export function RemixBrowser(_props: RemixBrowserProps): ReactElement {
     });
   }, [location]);
 
+  // eslint-disable-next-line react-hooks/rules-of-hooks
+  let fogOfWarAbortControllerRef = React.useRef<AbortController | undefined>();
+
+  // eslint-disable-next-line react-hooks/rules-of-hooks
+  React.useEffect(() => {
+    // Don't perform any prefetching if the app asked not to or the user has saveData enabled
+    if (discover !== "render" || navigator.connection?.saveData === true) {
+      return;
+    }
+
+    async function fetchPatches() {
+      if (fogOfWarAbortControllerRef.current) {
+        fogOfWarAbortControllerRef.current.abort();
+      }
+
+      let lazyPaths = getFogOfWarPaths();
+      if (lazyPaths.length === 0) {
+        return;
+      }
+
+      try {
+        fogOfWarAbortControllerRef.current = new AbortController();
+        await fetchAndApplyManifestPatches(
+          lazyPaths,
+          window.__remixManifest.version,
+          router.patchRoutes,
+          fogOfWarAbortControllerRef.current.signal
+        );
+      } catch (e) {
+        console.error("Failed to fetch manifest patches", e);
+      }
+    }
+
+    let debouncedFetchPatches = debounce(fetchPatches, 100);
+
+    function registerPath(path: string | null) {
+      if (!path || knownGoodPaths.has(path) || known404Paths.has(path)) return;
+      nextPaths.add(path);
+    }
+
+    let observer = new MutationObserver((records) => {
+      records.forEach((r) => {
+        if (r.type !== "childList") return;
+
+        r.addedNodes.forEach((node) => {
+          if (node.nodeType !== Node.ELEMENT_NODE) return;
+          let el = node as Element;
+          let links = Array.from(el.querySelectorAll("a[data-discover]"));
+          if (el.tagName === "A" && el.getAttribute("data-discover")) {
+            links.push(el);
+          }
+          links.forEach((el) => registerPath(el.getAttribute("href")));
+          debouncedFetchPatches();
+        });
+      });
+    });
```

Comment:
This logic detects newly rendered links on navigations and internal route state changes (i.e., click this button to show a link) and batches them up via the debounced fetch